File: merge_sort.py
"""Merge Sort Implementation in Python"""
import sys

def load_input(input_txt):
    """Loading input from .txt file to list"""
    with open(input_txt, 'r', encoding='utf-8') as f:
        data = [int(line.strip()) for line in f]
    return data

def merge_sort(input_array, start, end):
    """Recursive merge sort function that divides the input array and 
    calls the merge function to sort and merge the divided arrays."""
    if start < end:
        mid = (start + end) // 2
        merge_sort(input_array, start, mid)
        merge_sort(input_array, mid+1, end)
        merge(input_array, start, mid, end)

def merge(input_array, start, mid, end):
    """Merging function that takes two sorted subarrays and merges them 
    into a single sorted array."""
    len1 = mid - start + 1
    len2 = end - mid
    left_arr = [0] * len1
    right_arr = [0] * len2

    for i in range(len1):
        left_arr[i] = input_array[start + i]
    for j in range(len2):
        right_arr[j] = input_array[mid + 1 + j]

    i = 0
    j = 0
    k = start

    while i < len1 and j < len2:
        if left_arr[i] <= right_arr[j]:
            input_array[k] = left_arr[i]
            i += 1
        else:
            input_array[k] = right_arr[j]
            j += 1
        k += 1

    while i < len1:
        input_array[k] = left_arr[i]
        i += 1
        k += 1

    while j < len2:
        input_array[k] = right_arr[j]
        j += 1
        k += 1

def is_sorted(input_array):
    return all(input_array[i] <= input_array[i+1] for i in range(len(input_array)-1))

if __name__ == "__main__":
    input_string = sys.argv[1]
    n_loops = int(sys.argv[2])
    array = load_input(input_string)
    unsorted_array = array.copy()
    for _ in range(n_loops):
        merge_sort(array, 0, len(array)-1)
        array = unsorted_array.copy()
    if is_sorted(input_array=array):
        print("Sort successful")
    else:
        print("Sort failed")

File: mergeSort.c
// Mergesort in C
#include <stdio.h>
#include <stdlib.h>

void mergeSort(int *list, int start, int end);
void merge(int *list, int start, int mid, int end);
int is_sorted(int *list, int size);

int main (int argc, char *argv[]) {

    FILE *fp = fopen(argv[1], "r");
    if (fp == NULL) { 
        printf("Error opening file.\n");
        return 1;
    }

    // Find size of input list
    int listSize = 0;
    int temp;
    while (fscanf(fp, "%d", &temp) == 1) {
        listSize++;
    }

    // Rewind filepointer and then read all numbers into an array
    rewind(fp);
    int *list = (int *)malloc(listSize * sizeof(int));
    for (int i = 0; i < listSize; i++) {
        fscanf(fp, "%d", &list[i]);
    }
    // Sort the list using merge sort
    mergeSort(list, 0, listSize - 1);
    if(is_sorted(list, listSize)) {
        printf("Sort successful\n");
    } else {
        printf("Sort failed\n");
    }

    free(list);
    fclose(fp);
    return 0;
}

void mergeSort(int *list, int start, int end) {
    if (start < end) {
        int mid = start + (end - start) / 2;
        mergeSort(list, start, mid);
        mergeSort(list, mid + 1, end);
        merge(list, start, mid, end);
    }
}

void merge(int *list, int start, int mid, int end) {

    int len1 = mid - start + 1;
    int len2 = end - mid;
    int *leftArr = malloc(len1 * sizeof(int));
    int *rightArr = malloc(len2 * sizeof(int));

    for (int i = 0; i < len1; i++) {
        leftArr[i] = list[start + i];
    }
    for (int j = 0; j < len2; j++) {
        rightArr[j] = list[mid + 1 + j];
    }

    int i = 0, j = 0;
    int k = start;

    while (i < len1 && j < len2) {
        if (leftArr[i] <= rightArr[j]) {
            list[k] = leftArr[i];
            i++;
        } else {
            list[k] = rightArr[j];
            j++;
        }
        k++;
    }

    while (i < len1) {
        list[k] = leftArr[i];
        i++;
        k++;
    }

    while (j < len2) {
        list[k] = rightArr[j];
        j++;
        k++;
    }
}

int is_sorted(int *list, int size) {
    for (int i = 0; i < size - 1; i++) {
        if (list[i] > list[i + 1]) {
            return 0;
        }
    }
    return 1;
}